### DISCUSSION TITLE
RTSP Server

### DISCUSSION THREAD

**Discussion by RoySalisbury** [aaronwmorris/indi-allsky#1123] [Ideas]
Background:
When I was trying my hand at my own allsky camera awhile back, one of the tings I struggled with was making the timelapse video of the evening images.  The process I ended up with was taking the CURRENT image every second and saving it to a temp file.  No matter if it was a 20 second image or a 2 second image, I would make a copy every 1 second.  Then, on my hourly "timelapse" I would feed each of these images into ffmpeg to create a timelapse and set the FPS to 30.  This worked really well and I had a very nice timelapse.

Request:
I'm wondering if there is a way to do something similar but instead of doing it hourly off of temporary jpg images, you could do it as a "live" rtsp video.  The current image would be piped into ffmpeg from memory to simulate a 15/30 fps stream and then encoded into an .h264 live stream.

This could then be used to PUSH the stream into something like youtube for a live stream.

Currently I am doing something "similar" but using a 3rd party camera security package (BlueIris) to pullythe jpg once a second from the allsky website and then setting the video encoder to 1 FPS. I can then use BlueIris to push the stream to YouTube.  It works, but seems like a workaround that indi-allsky might be able to do itself, and not have all the encoding/decoding/reencoding to that degrades quality.

**Comment by aaronwmorris** [comment]
I am not sure about creating a live RTPS stream.  I am not saying it cannot be done, but I have never done it.  It might be interesting if ffmpeg could read images from a FIFO and stream that live.

I came up with this command as an example (just leaving this as a note for me later)
```
ffmpeg.exe -vcodec bmp -framerate 1 -i \\.\pipe\input_pipe -r 25 -c:v libx264 -preset ultrafast -tune zerolatency -pix_fmt yuv420p -f rtsp rtsp://wowza.server.com:1935/app/stream
```

I will take a longer look at this next week when I return from travel.

**Reply by RoySalisbury** [reply]
This is something similar to what I tried, but the issue I was running into and never went into deeper was that if you only push the image (via the pipe) as images are created, ffmpeg will time out because the frame rate can’t be maintained. 

I was only doing it once a second and telling ffmpeg I wanted a 30 fps stream. I think I would have had to actually push 30 fps via the pipe (even if they did not change). 

I may also try this again (it would be in C#, but the concept is the same). It would be a separate process that just reads a jpg file and pipes it into ffmpeg at 30 fps.  The piped image data would just be re-read from the jpg file every 1 second so it would get updates.  I don’t think a raspberry pi could do 30 fps, but perhaps 5 fps is possible. 

A python script might be able to do the same, I just don’t know python.

**Reply by RoySalisbury** [reply]
Quick update.  I did get this to work "sort of".  I wrote a simple program that takes a jpg file and sends it to a pipe at a fps of 30.   I then use the following ffmpeg command:

`ffmpeg.exe -hide_banner -y -r 30 -f image2pipe -s 1600x1080  -i \\.\pipe\test_pipe -c:v libx264 -preset ultrafast -tune zerolatency -pix_fmt yuv420p out.mp4`

This will then outpout the video.

I then tried pushing to youtube (my rtmp server in this case):

`ffmpeg.exe -hide_banner -y -r 30 -f image2pipe -s 1600x1080  -i \\.\pipe\test_pipe -c:v libx264 -preset ultrafast -tune zerolatency -pix_fmt yuv420p -f flv rtmp://a.rtmp.youtube.com/live2/<stream_key_removed>`

FFMPEG seems to think its sending the data, but youtube never seems to receive it (or its not in a format it can use).

So one step closer. :)

**Reply by RoySalisbury** [reply]
OK... A bit further...  I installed a local RTSP server for testing.

`docker run --rm -it --network=host bluenviron/mediamtx`

Then was able to run the ffmpg publisher
`
ffmpeg.exe -hide_banner -y -r 30 -f image2pipe -s 1600x1080  -i \\.\pipe\test_pipe -c:v libx264 -preset ultrafast -tune zerolatency  -vf "scale=1080:-2,setsar=1,format=yuv420p" -f rtsp rtsp://192.168.2.147:8554/mystream`

and then could view it using

`ffplay -rtsp_transport tcp rtsp://192.168.2.147:8554/mystream`

But it is a realtime stream now....

**Reply by RoySalisbury** [reply]
OK.. Finally got youtube to accept the stream.  You have to have an audio stream with it.  So I added a null aac stream.  

`ffmpeg.exe -hide_banner -y -f lavfi -i anullsrc  -r 30 -f image2pipe -s 1950x1310  -i \\.\pipe\test_pipe -c:v libx264 -preset ultrafast -tune zerolatency  -c:a aac -f flv rtmp://a.rtmp.youtube.com/live2/<key>`

I think I can downlevel the audio .. by default its 44100 Hz @ 128 kb/s ... Should bring that down to save bandwidth.

I also ned to figure out the output video size.  Even though the input is 1950x1310, its getting rendered @ a very low resolution (360p).

**Reply by RoySalisbury** [reply]
1080@30fps (8k mono audio)

`ffmpeg.exe -hide_banner -y -f lavfi -i anullsrc  -r 30 -f image2pipe -s 1950x1310  -i \\.\pipe\test_pipe -c:v libx264 -preset ultrafast -tune zerolatency -b:v 512k   -vf "scale=1920:-2,format=yuv420p" -c:a aac -b:a 8k -ac 1 -f flv rtmp://a.rtmp.youtube.com/<key>`
